TijanaSonjic/assignment2/Q2.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+#include <iostream>
+using namespace std;
+
+class Node {
+private:
+	Node* leftChild;
+	Node* rightChild;
+	int info;
+public:
+	Node(int info, Node* leftChild = nullptr, Node* rightChild = nullptr) {
+		this->info = info;
+		this->leftChild = leftChild;
+		this->rightChild = rightChild;
+	}
+
+	void distroy() {
+		if (leftChild) {
+			leftChild->distroy();
+			leftChild = nullptr;
+		}
+		if (rightChild) {
+			rightChild->distroy();
+			rightChild = nullptr;
+		}
+	}
+
+	~Node() {
+		distroy();
+	}
+
+	Node* insertLeftChild(Node* node) {
+		this->leftChild = node;
+
+		return this;
+	}
+
+	Node* insertRightChild(Node* node) {
+		this->rightChild = node;
+
+		return this;
+	}
+
+	int getInfo() {
+		return info;
+	}
+
+	Node* getLeftChild() {
+		return leftChild;
+	}
+
+	Node* getRightChild() {
+		return rightChild;
+	}
+};
+
+bool isInTheTree(Node* root, Node* node) {
+	if (!root)
+		return false;
+
+	if (root == node)
+		return true;
+
+	return isInTheTree(root->getLeftChild(), node) || isInTheTree(root->getRightChild(), node);
+}
+
+Node* findFirstCommonAncestor(Node* root, Node* node1, Node* node2) {
```

Comment:
If that is a helper function not intended to be called directly, there are several options to protect the users from using it:
1. Include it only in .cpp file and not in .h file (as an implementation detail)
2. If part of a class, declare it as private
3. Add a comment that clarifies this: `// This is a helper function, do not call directly. Call firstCommonAncestor instead.`
4. Give it a name that clarifies that this is a helper function, e.g. `firstCommonAncestorAux`.

I think in this context, 3 and 4 are good options. (No need to separate solutions to .h and .cpp in the assignments)

The point is to think of possible future users of this code, and think what we can do in order to make everyone use the code as we intended them to, making it hard to use it wrong. Calling this function directly may lead to unexpected results.

---

Also, this function assumes that both nodes are in the tree, so it is best to write this in a comment (to help the reader understand this function as an autonomous block).

Reply:
Actually findFirstCommonAncestor is a helper function and as you said this function assumes that both nodes are in the tree, so it is called from firstCommonAncestor function. Function firstCommonAncestor will check if any of function input arguments are null and if they are not null it will check if both nodes are in the tree before we call findFirstCommonAncestor which is the function with the main logic.  
In the future I will write comments like this one you have mentioned in third option, they are good practice and make code more readable. :)